### PR TITLE
IPv6 lab changes!

### PR DESCRIPTION
**Pull request by dougbtv**
Thanks @fepan

### CODE REVIEW

File: group_vars/all.yml
Comment (leifmadsen) [inline group_vars/all.yml:53]: Good start. I'm already going to modify this :)

File: group_vars/all.yml
Comment (leifmadsen) [inline group_vars/all.yml]: Hrmmm... I think we should call this `group` and not `role` to match Ansible naming conventions, unless you wanna go with something totally different. Perhaps something like.... `node_group` or `node_type`

File: group_vars/all.yml
Comment (leifmadsen) [inline group_vars/all.yml]: I don't really like how simple this is. Can it maybe something like `deploy_with_ipv6` or something else a bit more verbose?

File: inventory/examples/virthost/virthost-ipv6.inventory.yml
Comment (leifmadsen) [inline inventory/examples/virthost/virthost-ipv6.inventory.yml]: This would need to be updated to match the change requested above. `node_type` or something.

File: ipv6-lab.yml
Comment (leifmadsen) [inline ipv6-lab.yml]: `package`

File: ipv6-lab.yml
Comment (leifmadsen) [inline ipv6-lab.yml]: Can you break this out into a `handlers/` directory?

File: ipv6-lab.yml
Comment (leifmadsen) [inline ipv6-lab.yml]: Alternative approach which I found worked when I ran into a `poll` issue with Ansible:

```
    - name: Update the system                                                      
        yum:                                                                         
          update_cache: yes                                                          
          name: "*"                                                                  
          state: latest                                                              
        register: system_updates                                                     
                                                                                     
      - block:                                                                       
        - name: reboot system                                                        
          command: shutdown -r now "Ansible updates triggered"                       
          async: 0                                                                   
          poll: 0                                                                    
          ignore_errors: true                                                        
                                                                                     
        - name: waiting for server to come back                                      
          local_action: wait_for host={{ ansible_host | default(inventory_hostname,boolean=True) }} port=22 state='started' delay=5 timeout=300
          become: false                                                              
        when: system_updates.changed == "true"       
```

File: ipv6-lab.yml
Comment (leifmadsen) [reply ipv6-lab.yml]: Also, this might do what you think it does. Handlers aren't run until the end of the play execution, so if you want this to happen before moving to the next task, you need to use the approach I note below.

File: ipv6-lab.yml
Comment (leifmadsen) [inline ipv6-lab.yml:38]: This won't run until the end FYI

File: ipv6-lab.yml
Comment (leifmadsen) [inline ipv6-lab.yml]: delete comment blocks if you're not using them

File: ipv6-lab.yml
Comment (leifmadsen) [inline ipv6-lab.yml]: s/latest/present otherwise you're always updating things (potentially)

File: ipv6-lab.yml
Comment (leifmadsen) [inline ipv6-lab.yml]: `package`

File: ipv6-lab.yml
Comment (leifmadsen) [inline ipv6-lab.yml]: `present`

File: ipv6-lab.yml
Comment (leifmadsen) [inline ipv6-lab.yml:106]: Yea I kind of hate this :)

Make it a role and use ansible stuff to manage this

File: ipv6-lab.yml
Comment (leifmadsen) [inline ipv6-lab.yml:161]: `systemd` is probably what you want, not `service`

File: ipv6-lab.yml
Comment (leifmadsen) [inline ipv6-lab.yml]: fix this syntax to not be inline

```
- name: restart named
  systemd:
    name: named
    state: restart
```

File: ipv6-lab.yml
Comment (leifmadsen) [inline ipv6-lab.yml]: Not sure what this is accomplishing? Is this a test? If so, you should register and then run an assert against it.

File: roles/ipv6-setup/handlers/main.yml
Comment (leifmadsen) [inline roles/ipv6-setup/handlers/main.yml:4]: You could actually use this: http://docs.ansible.com/ansible/latest/net_interface_module.html

File: roles/ipv6-setup/handlers/main.yml
Comment (leifmadsen) [inline roles/ipv6-setup/handlers/main.yml:7]: What would be better is to poll and wait for the interface than a static pause value. Similar idea to what you were doing with waiting for a machine to reboot earlier.

File: roles/ipv6-setup/handlers/main.yml
Comment (leifmadsen) [inline roles/ipv6-setup/handlers/main.yml:16]: `systemd`

File: roles/ipv6-setup/tasks/main.yml
Comment (leifmadsen) [inline roles/ipv6-setup/tasks/main.yml]: `package`

File: roles/ipv6-setup/tasks/nodes.yml
Comment (leifmadsen) [inline roles/ipv6-setup/tasks/nodes.yml]: can you use the expanded form here?

File: roles/ipv6-setup/templates/ifcfg.j2
Comment (leifmadsen) [inline roles/ipv6-setup/templates/ifcfg.j2:1]: I wonder if just using Network Manager here actually makes this whole thing a lot simpler?

File: roles/kube-install/handlers/main.yml
Comment (leifmadsen) [inline roles/kube-install/handlers/main.yml:4]: `systemd`

File: roles/kube-install/tasks/binary_install.yml
Comment (leifmadsen) [inline roles/kube-install/tasks/binary_install.yml]: `package`

File: roles/kube-install/tasks/binary_install.yml
Comment (leifmadsen) [inline roles/kube-install/tasks/binary_install.yml:29]: Why can't we use `curl` or `get_url` here?

File: roles/kube-install/tasks/main.yml
Comment (leifmadsen) [inline roles/kube-install/tasks/main.yml:79]: Yep, hate this too :)

File: roles/kube-install/tasks/main.yml
Comment (leifmadsen) [inline roles/kube-install/tasks/main.yml]: `systemd`

File: roles/kubectl-proxy-systemd/tasks/main.yml
Comment (leifmadsen) [inline roles/kubectl-proxy-systemd/tasks/main.yml:5]: Why are we switching away from `get_url`?

File: roles/virthost-basics/tasks/main.yml
Comment (leifmadsen) [inline roles/virthost-basics/tasks/main.yml:18]: These are only needed when doing `with_ipv6` right?

File: roles/vm-spinup/templates/spinup.sh.j2
Comment (leifmadsen) [inline roles/vm-spinup/templates/spinup.sh.j2]: I don't think you need to indent here

File: roles/vm-spinup/templates/vms.local.j2
Comment (leifmadsen) [inline roles/vm-spinup/templates/vms.local.j2:21]: Don't do this :)  Only enable become when you need it, and not globally.

**Review by leifmadsen — CHANGES_REQUESTED** [review]
So there are a bunch of things I noted :)

File: group_vars/all.yml
Comment (fepan) [reply group_vars/all.yml]: This tag here isn't related to ansible, it indicates the node's role in the cluster being stood up. node_type would work too, but not sure that it matters.

File: group_vars/all.yml
Comment (fepan) [reply group_vars/all.yml]: I'm not sure that deploy_with_ipv6 conveys more information than just ipv6

File: ipv6-lab.yml
Comment (fepan) [reply ipv6-lab.yml]: will fix

File: ipv6-lab.yml
Comment (fepan) [reply ipv6-lab.yml:161]: I'd think service is what I want, if we are trying to use generic modules like package instead of yum.

File: ipv6-lab.yml
Comment (fepan) [reply ipv6-lab.yml]: Will move this to handlers file. 

This should work as expect. The handler is run at the end of play.

File: ipv6-lab.yml
Comment (fepan) [reply ipv6-lab.yml]: latest is intended. I was getting some conflict with existing packages not being up to date.

File: ipv6-lab.yml
Comment (fepan) [reply ipv6-lab.yml]: done

File: ipv6-lab.yml
Comment (fepan) [reply ipv6-lab.yml]: done

File: ipv6-lab.yml
Comment (fepan) [reply ipv6-lab.yml]: see above, latest is intended.

File: ipv6-lab.yml
Comment (fepan) [reply ipv6-lab.yml:106]: This is very specific to this node and the action is only performed once, not sure that we need to create a role.

File: ipv6-lab.yml
Comment (fepan) [reply ipv6-lab.yml]: done

File: ipv6-lab.yml
Comment (fepan) [reply ipv6-lab.yml]: good point, will fix.

File: roles/ipv6-setup/handlers/main.yml
Comment (fepan) [reply roles/ipv6-setup/handlers/main.yml:4]: I think this is only for network devices like cisco routers, do you have an example of this working for Linux?

File: roles/ipv6-setup/tasks/main.yml
Comment (fepan) [reply roles/ipv6-setup/tasks/main.yml]: done

File: roles/ipv6-setup/tasks/nodes.yml
Comment (fepan) [reply roles/ipv6-setup/tasks/nodes.yml]: done

File: roles/ipv6-setup/templates/ifcfg.j2
Comment (fepan) [reply roles/ipv6-setup/templates/ifcfg.j2:1]: yeah eventually network interfaces should be managed by nmcli.

File: roles/kube-install/tasks/binary_install.yml
Comment (fepan) [reply roles/kube-install/tasks/binary_install.yml]: done

File: roles/kube-install/tasks/binary_install.yml
Comment (fepan) [reply roles/kube-install/tasks/binary_install.yml:29]: get_url doesn't support ipv6 yet. curl works too, I just happened to use wget.

File: roles/kube-install/tasks/main.yml
Comment (fepan) [reply roles/kube-install/tasks/main.yml:79]: I think there should be a way to specify a version or point to binary from config. But for now this should be ok.

File: roles/kubectl-proxy-systemd/tasks/main.yml
Comment (fepan) [reply roles/kubectl-proxy-systemd/tasks/main.yml:5]: ipv6 support isn't quite there yet.

File: roles/vm-spinup/templates/spinup.sh.j2
Comment (fepan) [reply roles/vm-spinup/templates/spinup.sh.j2]: done

File: roles/virthost-basics/tasks/main.yml
Comment (dougbtv) [reply roles/virthost-basics/tasks/main.yml:18]: Somewhat. At least now these were necessary for the `virt_net` plays. So they might be otherwise useful in the future, at a fairly low cost of having the required packages on the virt host.

File: group_vars/all.yml
Comment (leifmadsen) [reply group_vars/all.yml]: I understand this isn't related to Ansible, but for readability of the code, I'd much prefer `node_type` here over an Ansible naming convention since it's a simple change to make.

File: group_vars/all.yml
Comment (leifmadsen) [reply group_vars/all.yml]: The point being that I think just using `ipv6` plain isn't a good variable. I suspect there will be other `ipv6` related variables, so those would likely end up being `ipv6_foo` etc, so starting with something like `ipv6_enabled: false` would be a much better variable name.

File: group_vars/all.yml
Comment (fepan) [reply group_vars/all.yml]: Sure, will change to node_type

File: ipv6-lab.yml
Comment (leifmadsen) [reply ipv6-lab.yml]: If you're installing the Development tools here fresh on a new VM, then use of `latest` shouldn't be necessary. This will fail ansible linting. If you really need to run an upgrade, then you should move `latest` into a package upgrade playbook, or make this a variable, with default to `present`.

File: ipv6-lab.yml
Comment (leifmadsen) [reply ipv6-lab.yml:106]: The use of inline `shell` like this is incredibly ugly, and not very portable or idempotent.

File: ipv6-lab.yml
Comment (leifmadsen) [reply ipv6-lab.yml:161]: CentOS 7 is the target here though, which uses systemd, not initd.

File: roles/kube-install/tasks/main.yml
Comment (leifmadsen) [reply roles/kube-install/tasks/main.yml:79]: I meant the whole expanded `shell` thing.

File: roles/kubectl-proxy-systemd/tasks/main.yml
Comment (leifmadsen) [reply roles/kubectl-proxy-systemd/tasks/main.yml:5]: As mentioned in IRC, don't rip out all the known working `get_url`. Wrap it in a `when` clause to skip when using `ipv6_enabled` and execute your `wget` version when using `ipv6_enabled`. That way we don't have to remember to go back and re-convert back to `get_url` when it does eventually work.

(I know @dougbtv is testing this now, so ideally we can just drop all this `wget` stuff before we even merge.)

File: ipv6-lab.yml
Comment (fepan) [reply ipv6-lab.yml:161]: if Centos 7 is the target, wouldn't we want to use yum and not package?

File: roles/vm-spinup/templates/vms.local.j2
Comment (leifmadsen) [reply roles/vm-spinup/templates/vms.local.j2:21]: CC: @dougbtv as I think you're testing this now?

File: roles/kubectl-proxy-systemd/tasks/main.yml
Comment (leifmadsen) [reply roles/kubectl-proxy-systemd/tasks/main.yml:5]: Not sure what is more clean, but if you installed EPEL, you could just install `jq` from the packaging system.

File: group_vars/all.yml
Comment (fepan) [reply group_vars/all.yml]: will change to ipve_enabled

**Comment by leifmadsen** [comment]
Do not merge this branch. We'll have to merge in `feature/ipv6` as it seems cleaner and rebases a lot nicer.

**Comment by leifmadsen** [comment]
Closing in favour of a different branch

**Pull request closed without merging**
